fancy-calling-conventions.md

```diff
@@ -0,0 +1,175 @@
+
+### 8.  args inside class method (needs lists and iterators)
+```
+class Foo(object):
+	bar : int[] = []
+	def load(self: Foo, * args : int) -> None:
+		for arg in args:
+			self.bar.append(arg)
+foo : Foo = None
+foo = Foo()
+foo.load(1, 2, 3, 4, 5)
+print(foo.bar) // [1, 2, 3, 4, 5]
+```
+This program tests the functionality of arbitrary arguments in calls to methods within a class. Similar to arbitrary arguments used in function definitions, we expect that calls to methods with arbitrary parameters should be able to take any number of arguments with the specified type. In this case, the load method in Foo should take an arbitrary number of integer arguments.
+
+### 9.  Default and args
+```
+def foo(base : int = 0, * args : int) -> int:
+	for arg in args:
+		base += arg
+return base
+foo(1, 2, 3, 4, 5) // should return 15
+```
+Combines default parameters with arbitrary arguments. We ensure that regular parameters and parameters with default values are defined before the arbitrary arguments.
+
+### 10.  args and kwargs (needs strings, dicts, lists, iterators, destructors?)
+```
+def foo(* args, ** kwargs) -> None:
+	for arg in args:
+		print(arg)
+	for key, value in kwargs.items():
+		print(key)
+		print(value)
+foo(1, 2, 3, a = 4, b = 5, c = 6)
+
+“””
+1
+2
+3
+a
+4
+b
+5
+c
+6
+“””
+```
+Combining arbitrary arguments with keyword arguments. The convention follows that arbitrary arguments are defined before the keyword arguments. For this program to work, we treat arbitrary arguments as a list and keyword arguments as a dictionary. Finally, there would need to be support for iterators for the program to walk through these list and dictionary structures.
+
+## Tests:
+
+### Default arguments
+For default arguments, we will type check the literals defined at the function signature to make sure it matches the type of parameter. For function calls and method calls, we have to check the number of parameters passed in. The number of parameters passed in must be at least greater than the number of parameters without default values and less than the number of total parameters defined in the function signature. Then, those values in the function/method calls must be type checked.
+
+### Arbitrary arguments
+For arbitrary arguments, we need to make sure that all of the arguments passed in matches with the type defined at the parameter.
+
+### Keyword arguments
+For testing keyword arguments, we will switch the argument positions but ensure that the output of the function stays the same.
+
+### Regular arguments
+To ensure the basic functionality of method calls and function calls are still intact, we will write the exact program without the new calling conventions and make sure the outputs are correct.
+
+## New AST forms:
+
+To account for parameters with default arguments, we would need to include a new pattern for the `Parameter` type in `ast.ts`. The new pattern should include a field for the default value.
```

Comment:
Assume that the `Type` for callable is something like `Callable[[int, bool], int]` in Python and below in AST
``` typescript
export type Type = ...
    | {tag: "callable", args: Array<Type>, ret: Type}
```
How should it be changed to enable default arguments?

Reply:
Default arguments shouldn't change for callables.
For args and kwargs, it can change.
I feel like args and kwargs can be optional... args will just be another `Array<Type>`, and kwargs can be `Map<string, Type>`... Does that make sense?

I envision something like:
```
{tag: "callable", args: Array<Type>, * args: Array<Type>, ** kwargs: Map<string, Type>, ret: Type}
```

Reply:
> Default arguments shouldn't change for callables.

I don't really understand this. For example, the `foo` in  `def foo(int:x = 1) -> int:` (func body omitted) should be of type `Callable[[int], int]`, but it should also be of type (able to be conformed/derived as) `Callable[[], int]`. There must be some changes to deal with that problem.

Reply:
For your example, I think `Callable[[int], int]` would be the correct typing. 
Do you foresee problems occuring in the typecheck? That's the only place I think problems could arise (specifically checking the arguments against the predefined parameter types).

Reply:
Well, if you give the type `Callable[[int], int]` to `def foo(int:x = 1) -> int:`, I guess you assumeed that `Callable[[int], int]` can be uses as an callee that requires **one or less** arguments. In that case, the following would be considered as type correct while it is actually not.

``` python
id: Callable[[int], int] = None

def f(x:int) -> int:
  return x

id = f
id()  # should raise an error during type checking
```

Reply:
Yeah, that would not be correct.
If it were something like this though, it would be correct:
```
def f(x:int = 1) -> int:
    return x
id = f
id()
```
Here, an argument is not supplied to `id`, so the definition of `f` will use the default argument.

Reply:
@DDEle 

Hey there, sorry for the late reply. Just want to give you guys some updates on what me, @danielytan and @DavidLiu0304 are doing.
Our work hinges on knowing what the type environment is. Therefore, much of our work for default arguments is done in type check. For default arguments, if the user does not supply arguments but default arguments are present, we insert the default arguments in the call expression using the decorated AST technique.
(We are also using this similar technique to implement keyword arguments.)

Now, how do I think this will affect your work on closures? Suppose we have something like this:
```
def f(x : int, y : int = 2) -> int:
    return x + y
g = f
print(g(1)) # prints 3 
print(g(1, 3)) # prints 4
```
Just like as if the `y` parameter is required, `f` in the above snippet should still be of type `Callable[[int, int], int]`, even though we can give callee less than the number of arguments defined in the function because of default functions. In our type check, we will see a call expression and look up its corresponding type information. In the first call, we see that we need default arguments and thus we edit the AST to insert the default argument. In the second call, since we don't need additional default arguments, nothing changes in the AST.
The advantage to this approach is that it doesn't change the code in `compiler.ts`, because once the default arguments are inserted as needed, the function call itself is no longer anything special.
Let me know if you have any questions. Thanks.

Reply:
@xiexr151e Thanks for updating. I'm still worried that if we can combine the default parameter with the first level functions. Considering this example
``` python
def fd2(x : int, y : int = 2) -> int:
    return x + y

def fd3(x : int, y : int = 3) -> int:
    return x + y

g = None
if (a_nontrivial_expression_of_boolean):
    g = fd2
else:
    g = fd3

print(g(1)) # prints 3 or 4?
```